### PR TITLE
[deckhouse-config] fix: ignore ModuleConfig resources with invalid spec.settings

### PR DESCRIPTION
**Pull request by diafour**
## Description

- Split ConfigValidator.Validate to three methods: ValidateCR, ConvertToLatest, Validate.
- Use ConvertToLatest in initial creation of ModuleConfig
- Ignore ModuleConfig with invalid spec.settings

Giving this invalid "openvpn" section in the ConfigMap/deckhouse:

```
apiVersion: v1
kind: ConfigMap
metadata:
  name: deckhouse
  namespace: d8-system
data:
  global: |
    ...
  openvpn: |
    pushToClientRoutes:
        - 10.111.0.0/16
        - 10.222.0.0/16
    pushToClientSearchDomains:
        - pew.mad
  openvpnEnabled: "true"
```

ModuleConfig/openvpn will be created with following status:

## Why do we need it, and what problem does it solve?

A continuation of #1729: update to release with #1729 may fail on outdated cm/deckhouse. Recovery from this fail is non-automatic.

It also fix possible problems with conversions at start (see Corner cases 1 and 5 in #3097)

## What is the expected result?

Deckhouse not fail on cm/deckhouse with outdated invalid module sections.

## Checklist
- [ ] The code is covered by unit tests.
- [ ] e2e tests passed.
- [ ] Documentation updated according to the changes.
- [ ] Changes were tested in the Kubernetes cluster manually.

## Changelog entries

```changes
section: deckhouse-config
type: fix
summary: Ignore ModuleConfig resources with invalid spec.settings.
impact_level: low
```

### CODE REVIEW

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @z9r5. I've started the workflow for label 'e2e/run/aws'...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @z9r5. I've started the workflow for label 'e2e/run/aws'...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @z9r5. I've started the workflow for label 'e2e/run/gcp'...

File: go_lib/deckhouse-config/validator.go [hunk 1]
```diff
@@ -46,74 +46,139 @@ type ValuesValidator interface {
...
 	}
 
-	// Settings are present, check version using conversion chain.
+	// Check if there is registered conversion for the version and if the version is the latest.
 	chain := conversion.Registry().Chain(cfg.GetName())
 	latestVer := chain.LatestVersion()
 
+	// Check if version is unknown.
 	if !chain.IsKnownVersion(cfg.Spec.Version) {
 		previousVersions := concatIntList(chain.PreviousVersionsList())
```
Comment (shvgn) [inline go_lib/deckhouse-config/validator.go:108]: To outline the goal of the variable I'd call `previousVersions` to `previousVersionsMessage` or `prevVersionsMessage` .

File: go_lib/deckhouse-config/validator.go [hunk 2]
```diff
@@ -46,74 +46,139 @@ type ValuesValidator interface {
...
+		errs = append(errs, v.ConversionError)
+	}
+	if v.ValidationError != "" {
+		errs = append(errs, v.ValidationError)
+	}
+	return strings.Join(errs, ", ")
+}
+
+// ValidateCR checks if ModuleConfig resource is well-formed.
+func (c *ConfigValidator) ValidateCR(cfg *d8cfg_v1alpha1.ModuleConfig) ValidationResult {
```
Comment (shvgn) [inline go_lib/deckhouse-config/validator.go:81]: Let's call this method `ValidateModuleConfig` ?

Comment (shvgn) [reply go_lib/deckhouse-config/validator.go:81]: Nit: Instead of `ValidationResult`, it could return `error` that `ValidationResult` generates. And then Everything else can be deduced without `ValidationResult`, for example non-nil spec.settings.

File: go_lib/deckhouse-config/validator.go [hunk 3]
```diff
@@ -46,74 +46,139 @@ type ValuesValidator interface {
...
+	// Check if version is unknown.
 	if !chain.IsKnownVersion(cfg.Spec.Version) {
 		previousVersions := concatIntList(chain.PreviousVersionsList())
 		if previousVersions != "" {
 			previousVersions = fmt.Sprintf(", or one of previous versions: %s", previousVersions)
 		}
 
-		return result, fmt.Errorf("spec.version=%d is invalid. Use latest version %d%s", cfg.Spec.Version, latestVer, previousVersions)
+		msg := fmt.Sprintf("spec.version=%d is unsupported. Use latest version %d%s", cfg.Spec.Version, latestVer, previousVersions)
+		if result.HasVersionedSettings {
```
Comment (shvgn) [inline go_lib/deckhouse-config/validator.go:114]: On line 99 you just set effectively 
```go
result.HasVersionedSettings = cfg.Spec.Settings != nil
```

And here you check the same fact.

Comment (shvgn) [reply go_lib/deckhouse-config/validator.go:114]: In general I'd not distinguish empty settings from any other state. Why would we want to give it special attention?

File: go_lib/deckhouse-config/validator.go [hunk 4]
```diff
@@ -46,74 +46,139 @@ type ValuesValidator interface {
...
+	}
+	result.ValidateCRWarning = versionMsg
+	return result
+}
+
+// ConvertToLatest checks if ModuleConfig resource is well-formed and runs conversions for spec.settings is needed.
+func (c *ConfigValidator) ConvertToLatest(cfg *d8cfg_v1alpha1.ModuleConfig) ValidationResult {
+	result := c.ValidateCR(cfg)
+	if result.HasError() || !result.HasVersionedSettings {
+		return result
```
Comment (shvgn) [inline go_lib/deckhouse-config/validator.go:137]: Nitpicking

This code would be cleaner this way:

```go
if cfg.Settings == nil {
  return nil 
}
err := := c.ValidateCR(cfg)
if err != nil {
  return err
}
```

Seems, the ValidationResult copies the state of MC and you now track two states of the same data (MC + ValidationResult) instead of one. But what you are trying to do is to get an error, and the knowledge about non-empty settings and changed version. The latter two are easily accessible without ValidationResult. 

If we are to throw away ValidationResult, the code will remain responsible to combine multiple errors into one. [multierr](https://pkg.go.dev/github.com/uber-go/multierr#Combine) package must be useful here.

Additionally, we can wrap errors if we need to split the processing of them. But seems is it not the case.

Additionally, I think currently it is not idiomatic go code.

**Review by shvgn — CHANGES_REQUESTED** [review]

File: global-hooks/deckhouse-config/startup_sync_test.go [hunk 5]
```diff
@@ -161,7 +161,7 @@ unknownModule: |
 			})
 		})
 
-		Context("giving invalid ConfigMap", func() {
+		Context("giving invalid generated ConfigMap", func() {
 			BeforeEach(func() {
 				// Emulate migrated Deployment/deckhouse.
 				_ = os.Setenv("ADDON_OPERATOR_CONFIG_MAP", d8config.GeneratedConfigMapName)
```
[the remark this answers is not in the document]
Comment (diafour) [reply global-hooks/deckhouse-config/startup_sync_test.go:167]: Done

File: global-hooks/deckhouse-config/startup_sync.go [hunk 6]
```diff
@@ -202,19 +209,26 @@ func modifyDeckhouseDeploymentToUseGeneratedConfigMap(patchCollector *object_pat
 	patchCollector.Filter(modify, "apps/v1", "Deployment", d8config.DeckhouseNS, "deckhouse")
 }
 
+// syncModuleConfigs updates generated ConfigMap using ModuleConfig resources.
 func syncModuleConfigs(input *go_hook.HookInput, generatedCM *v1.ConfigMap, allConfigs []*d8cfg_v1alpha1.ModuleConfig) error {
+	properCfgs := make([]*d8cfg_v1alpha1.ModuleConfig, 0)
```
[the remark this answers is not in the document]
Comment (diafour) [reply global-hooks/deckhouse-config/startup_sync.go:214]: I've put it to https://github.com/deckhouse/deckhouse/issues/3097

File: global-hooks/deckhouse-config/startup_sync.go [hunk 7]
```diff
@@ -143,15 +143,22 @@ func createInitialModuleConfigs(input *go_hook.HookInput, cmData map[string]stri
...
 		}
+		// Update spec.settings to converted settings.
 		if res.IsConverted {
 			cfg.Spec.Settings = res.Settings
 			cfg.Spec.Version = res.Version
 		}
+		properCfgs = append(properCfgs, cfg)
 	}
 
 	input.LogEntry.Infof("Create %d ModuleConfig objects", len(configs))
```
[the remark this answers is not in the document]
Comment (diafour) [reply global-hooks/deckhouse-config/startup_sync.go:164]: Done

File: go_lib/deckhouse-config/validator.go (hunk 1 above)
Comment (diafour) [reply go_lib/deckhouse-config/validator.go:108]: Done

File: go_lib/deckhouse-config/validator.go (hunk 2 above)
Comment (diafour) [reply go_lib/deckhouse-config/validator.go:81]: Discussed. Decision is to make refactoring later.
Use error wrapping and use v1alpha1.ModuleConfig object instead of additional type.

File: go_lib/deckhouse-config/validator.go (hunk 3 above)
Comment (diafour) [reply go_lib/deckhouse-config/validator.go:114]: I've removed this public field, it is now a helper func.

File: go_lib/deckhouse-config/validator.go (hunk 4 above)
Comment (diafour) [reply go_lib/deckhouse-config/validator.go:137]: See https://github.com/deckhouse/deckhouse/pull/3125#discussion_r1032373001

File: go_lib/deckhouse-config/validator.go [hunk 8]
```diff
@@ -46,74 +46,139 @@ type ValuesValidator interface {
...
+		// Resource is valid without spec.version and spec.settings.
+		return result
 	}
-	if cfg.Spec.Settings == nil && cfg.Spec.Version > 0 {
-		return result, fmt.Errorf("spec.version is forbidden when settings are not specified")
+
+	// Can run conversions and validations if spec.version and spec.settings are specified.
+	if cfg.Spec.Settings == nil {
+		// Warn about spec.version without spec.settings.
+		result.ValidateCRWarning = "spec.version is redundant without spec.settings"
```
[the remark this answers is not in the document]
Comment (diafour) [reply go_lib/deckhouse-config/validator.go:96]: Discussed:
1. The version alone is not a valid case, but the contract is to ignore it and display warning in the status.
This contract may be distractive and we need to make it more clearer in the future.
2. Empty spec.settings and absent spec.settings should behave equaly.

File: global-hooks/deckhouse-config/startup_sync.go [hunk 9]
```diff
@@ -161,8 +161,8 @@ func createInitialModuleConfigs(input *go_hook.HookInput, cmData map[string]stri
 		properCfgs = append(properCfgs, cfg)
 	}
 
-	input.LogEntry.Infof("Create %d ModuleConfig objects", len(configs))
 	for _, cfg := range configs {
+		input.LogEntry.Infof("Create ModuleConfig/%s", cfg.GetName())
```
Comment (shvgn) [inline global-hooks/deckhouse-config/startup_sync.go:165]: "Create" is imperative. Maybe 'Creating'? 'Created'?

File: go_lib/deckhouse-config/validator.go [hunk 10]
```diff
@@ -93,30 +77,28 @@ func (c *ConfigValidator) ValidateCR(cfg *d8cfg_v1alpha1.ModuleConfig) Validatio
 	// Can run conversions and validations if spec.version and spec.settings are specified.
 	if cfg.Spec.Settings == nil {
 		// Warn about spec.version without spec.settings.
-		result.ValidateCRWarning = "spec.version is redundant without spec.settings"
-	} else {
-		// Resource has both spec.settings and spec.version — it is ok to convert and validate with OpenAPI schema.
-		result.HasVersionedSettings = true
+		result.Warning = "spec.version is redundant without spec.settings"
```
Comment (shvgn) [inline go_lib/deckhouse-config/validator.go:80]: "spec.version has no effect without spec.settings, defaults from the latest config version will be applied"

**Review by shvgn — APPROVED** [review]

File: global-hooks/deckhouse-config/startup_sync.go [hunk 11]
```diff
@@ -143,24 +143,31 @@ func createInitialModuleConfigs(input *go_hook.HookInput, cmData map[string]stri
...
 		if res.IsConverted {
 			cfg.Spec.Settings = res.Settings
 			cfg.Spec.Version = res.Version
 		}
+		properCfgs = append(properCfgs, cfg)
 	}
 
-	input.LogEntry.Infof("Create %d ModuleConfig objects", len(configs))
 	for _, cfg := range configs {
+		input.LogEntry.Infof("Create ModuleConfig/%s", cfg.GetName())
```
Comment (diafour) [inline global-hooks/deckhouse-config/startup_sync.go:165]:
```suggestion
		input.LogEntry.Infof("Creating ModuleConfig/%s", cfg.GetName())
```

File: go_lib/deckhouse-config/validator.go [hunk 12]
```diff
@@ -46,74 +46,123 @@ type ValuesValidator interface {
...
+		// Resource is valid without spec.version and spec.settings.
+		return result
 	}
-	if cfg.Spec.Settings == nil && cfg.Spec.Version > 0 {
-		return result, fmt.Errorf("spec.version is forbidden when settings are not specified")
+
+	// Can run conversions and validations if spec.version and spec.settings are specified.
+	if cfg.Spec.Settings == nil {
+		// Warn about spec.version without spec.settings.
+		result.Warning = "spec.version is redundant without spec.settings"
```
Comment (diafour) [inline go_lib/deckhouse-config/validator.go:80]:
```suggestion
		result.Warning = "spec.version has no effect without spec.settings, defaults from the latest version of settings schema will be applied"
```

**Review by shvgn — APPROVED** [review]

**Review by yalosev — APPROVED** [review]

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @yalosev. I've started the workflow for label 'e2e/run/yandex-cloud'...
